exonum_derive/src/lib.rs

```diff
@@ -0,0 +1,302 @@
+// limitations under the License.
+
+#![recursion_limit = "256"]
+
+extern crate proc_macro;
+extern crate proc_macro2;
+
+extern crate syn;
+#[macro_use]
+extern crate quote;
+
+use proc_macro::TokenStream;
+use proc_macro2::{Ident, Span};
+use syn::{Data, DeriveInput, Lit, Meta, NestedMeta, Path};
+
+use std::env;
+
+#[proc_macro_derive(ProtobufConvert, attributes(protobuf_convert))]
+pub fn protobuf_convert_derive(input: TokenStream) -> TokenStream {
+    let input: DeriveInput = syn::parse(input).unwrap();
+
+    let name = input.ident;
+    let proto_struct_name = {
+        input
+            .attrs
+            .iter()
+            .find_map(|attr| {
+                let meta = attr.parse_meta().ok()?;
+                if meta.name() != "protobuf_convert" {
+                    return None;
+                }
+                let list = match meta {
+                    Meta::List(x) => x,
+                    _ => panic!("protobuf_convert attribute expects one argument"),
+                };
+                let name: Path = match list.nested.iter().next().expect("h") {
+                    NestedMeta::Literal(Lit::Str(lit_str)) => lit_str
+                        .parse()
+                        .expect("protobuf_convert argument should be valid type path"),
+                    _ => panic!("protobuf_convert argument should be string"),
+                };
+                Some(name)
+            }).expect("protobuf_convert attribute not set")
+    };
+    let mod_name = Ident::new(&format!("pb_convert_impl_{}", name), Span::call_site());
+    let data = match input.data {
+        Data::Struct(x) => x,
+        _ => panic!("Protobuf convert can be derived for structs only"),
+    };
+
+    let field_names = data
+        .fields
+        .iter()
+        .map(|f| f.ident.clone().unwrap())
+        .collect::<Vec<_>>();
+
+    let to_pb_fn = {
+        let names1 = field_names.clone();
+        let setters = names1
+            .iter()
+            .map(|i| Ident::new(&format!("set_{}", i), Span::call_site()));
+        let names2 = field_names.clone();
+        quote! {
+            fn to_pb(&self) -> Self::ProtoStruct {
+                let mut msg = Self::ProtoStruct::new();
+                #( msg.#setters(self.#names2.to_pb().into()); )*
+                msg
+            }
+        }
+    };
+
+    let from_pb_fn = {
+        let names1 = field_names.clone();
+        let getters = names1
+            .iter()
+            .map(|i| Ident::new(&format!("get_{}", i), Span::call_site()));
+        let names2 = field_names.clone();
+        quote! {
+            fn from_pb(pb: Self::ProtoStruct) -> Result<Self, ()> {
+              Ok(Self {
+               #( #names2: ProtobufConvert::from_pb(pb.#getters().to_owned())?, )*
+              })
+            }
+        }
+    };
+
+    let cr = if env::var("CARGO_PKG_NAME").unwrap() == "exonum" {
+        quote!(crate)
+    } else {
+        quote!(exonum)
+    };
```

Comment:
Isn't there a better way than special-casing `exonum`? It looks like a code smell to me.

Reply:
Its one of the best for current rust, but it will be changes in future when you would have ability to write something like `export exonum as self` in exonum crate so you could write code as if exonum is external crate.
Before you could use `pub mod exonum{ pub use super::* } in crate root to do this, but it does not work now.